### PR TITLE
Ongoing OIDC work: provider cache and auth method Lookup and List repo methods

### PR DESCRIPTION
**Pull request by jimlambrt**
The PR adds:
1. An oidc.Provider cache, since we need to cache the JWKs used to verify ID Tokens.  Before getting a provider from the cache, we will retrieve the AuthMethod from the DB so we can determine if the provider configuration has been updated.

2. repo.LookupAuthMethod(...) and repo.ListAuthMethods(...) 

@s-christoff, @louisruch, @talanknight : I added you as optional reviewers.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -536,12 +536,14 @@ github.com/grpc-ecosystem/grpc-gateway v1.9.5/go.mod h1:vNeuVxBJEsws4ogUvrchl83t
 github.com/grpc-ecosystem/grpc-gateway/v2 v2.1.1-0.20210209144642-519e7f45f48e h1:9BVX/Q2UP8rvj1HjR7O260MCykUXjXXCEJEvDcY/GLI=
 github.com/grpc-ecosystem/grpc-gateway/v2 v2.1.1-0.20210209144642-519e7f45f48e/go.mod h1:gRq9gZWcIFvz68EgWqy2qQpRbmtn5j2qLZ4zHjqiLpg=
 github.com/hailocab/go-hostpool v0.0.0-20160125115350-e80d13ce29ed/go.mod h1:tMWxXQ9wFIaZeTI9F+hmhFiGpFmhOHzyShyFUhRm0H4=
-github.com/hashicorp/cap v0.0.0-20210204173447-5fcddadbf7c7 h1:6OHvaQs9ys66bR1yqHuoI231JAoalgGgxeqzQuVOfX0=
-github.com/hashicorp/cap v0.0.0-20210204173447-5fcddadbf7c7/go.mod h1:tIk5rB1nihW5+9bZjI7xlc8LGw8FYfiFMKOpHPbWgug=
 github.com/hashicorp/boundary/api v0.0.5 h1:u6gkEcQU2rZyE2XEGln/h2/cKxGRtaeJFP/xLSpQr7E=
 github.com/hashicorp/boundary/api v0.0.5/go.mod h1:7UdSNnUHkrvEn3HrD6hQYO/HtzpnYt6u6UDiCfLWn1o=
 github.com/hashicorp/boundary/sdk v0.0.3 h1:xCj1LLvoLkJ6qP6oyn9k5uVHn8N40TGRG+iw7GdDmmU=
 github.com/hashicorp/boundary/sdk v0.0.3/go.mod h1:hu1j0/j9FY+G/vhl+6zoT18dCFs+wzeyTI+EuTHg4OY=
+github.com/hashicorp/cap v0.0.0-20210204173447-5fcddadbf7c7 h1:6OHvaQs9ys66bR1yqHuoI231JAoalgGgxeqzQuVOfX0=
```
Comment (louisruch) [inline go.sum:543]: `go mod tidy`

File: internal/auth/oidc/auth_method.go [hunk 2]
```diff
@@ -197,3 +198,29 @@ func (a *AuthMethod) hmacClientSecret(cipher wrapping.Wrapper) error {
 	a.ClientSecretHmac = base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
 	return nil
 }
+
+// isComplete() checks the auth method to see if it has all the required
+// components of a complete/valid oidc auth method.
+func (am *AuthMethod) isComplete() error {
+	const op = "oidc.isComplete"
```
Comment (louisruch) [inline internal/auth/oidc/auth_method.go:205]:
```suggestion
	const op = "oidc.(AuthMethod).isComplete"
```

File: internal/auth/oidc/provider.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+// Before returning a cached oidc.Provider, getProvider ensures that the
+// AuthMethod data used for the oidc.Provider's configuration hasn't changed
+// since it was cached. This is necessary because another controller could
+// update the AuthMethod in the DB, which would require changing the
+// configuration of the cached provider.
+//
+// getProvider will update the providerCache with the new AuthMethod, if it
+// determines the provider's configuration has been updated in the DB.
+func (c *providers) getProvider(ctx context.Context, currentFromDb *AuthMethod) (*oidc.Provider, error) {
+	const op = "oidc.(Repository).getProvider"
```
Comment (louisruch) [inline internal/auth/oidc/provider.go:57]:
```suggestion
	const op = "oidc.(providers).getProvider"
```

File: internal/auth/oidc/provider.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	if ok {
+		cachedHash, err := p.ConfigHash()
+		if err != nil {
+			return nil, errors.Wrap(err, op, errors.WithMsg("unable to hash provider from cache"))
+		}
+		storedHash, err := storedProvider.ConfigHash()
+		if err != nil {
+			return nil, errors.Wrap(err, op, errors.WithMsg("unable to hash provider from database"))
+		}
+		switch cachedHash == storedHash {
```
Comment (louisruch) [inline internal/auth/oidc/provider.go:74]: I love seeing the commitment to using a switch!

**Review by louisruch — COMMENTED** [review]
Couple of small nits otherwise this looks great

File: internal/auth/oidc/provider.go (hunk 4 above)
Comment (jimlambrt) [reply internal/auth/oidc/provider.go:74]: I'm really trying to do this more consistently when it's an "or" condition I'm testing for.  ifs == ands and switch == ors

File: go.sum (hunk 1 above)
Comment (jimlambrt) [reply go.sum]: ty!

**Comment by jimlambrt** [comment]
I've updated this PR to include:
-  repo.DeleteAuthMethod(...) and repo.CreateAuthMethod(...)
- it removes the Stopping state, based on our decision last week. 

Making these additions to this PR will unblock work on the OIDC feature.

File: internal/auth/oidc/testing.go [hunk 5]
```diff
@@ -88,10 +96,37 @@ func TestAuthMethod(
...
+	// sort all the embedded value objects...
+	for _, am := range methods {
+		sort.Slice(am.SigningAlgs, func(a, b int) bool {
+			return am.SigningAlgs[a] < am.SigningAlgs[b]
+		})
+		sort.Slice(am.AudClaims, func(a, b int) bool {
+			return am.AudClaims[a] < am.AudClaims[b]
+		})
+		sort.Slice(am.CallbackUrls, func(a, b int) bool {
+			return am.AudClaims[a] < am.AudClaims[b]
```
Comment (louisruch) [inline internal/auth/oidc/testing.go:122]:
```suggestion
			return am.CallbackUrls[a] < am.CallbackUrls[b]
```

**Review by louisruch — COMMENTED** [review]
Looks great, thank you for the detailed godoc comments on the unimplemented functions!  I have had experience with a dev writing a bunch of `panic("TODO")` and then left the company with no one having a clear idea of what needed to be done

File: internal/auth/oidc/repository_auth_method_create.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		return nil, errors.Wrap(err, op)
+	}
+
+	var returnedAuthMethod *AuthMethod
+	_, err = r.writer.DoTx(
+		ctx,
+		db.StdRetryCnt,
+		db.ExpBackoff{},
+		func(_ db.Reader, w db.Writer) error {
+			msgs := make([]*oplog.Message, 0, 4)
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/auth/oidc/repository_auth_method_create.go:65]: there are msgs appended as the function continues... for a total of 5, so this was off by 1, which I've fixed.

File: internal/auth/oidc/repository_auth_method_delete.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		return db.NoRowsAffected, errors.Wrap(err, op)
+	}
+	if am == nil {
+		// already deleted and this is not an error.
+		return db.NoRowsAffected, nil
+	}
+
+	oplogWrapper, err := r.kms.GetWrapper(ctx, am.ScopeId, kms.KeyPurposeOplog)
+	if err != nil {
+		return db.NoRowsAffected, errors.Wrap(err, op, errors.WithCode(errors.Encrypt), errors.WithMsg("unable to get oplog wrapper"))
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/auth/oidc/repository_auth_method_delete.go:32]: ty!

File: internal/auth/oidc/repository_auth_method_create.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (louisruch) [reply internal/auth/oidc/repository_auth_method_create.go]: Each append is appending a slice which could be of any length, unless the common case here is 1?

**Review by louisruch — APPROVED** [review]
I updated my question re the size of 4, sorry if I was not explicit as to why I didn't believe it would hit 4 originally.  But regardless of that its a minor nit this LGTM

File: internal/auth/oidc/repository_auth_method_create.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+)
+
+// CreateAuthMethod creates m (*AuthMethod) in the repo along with its
```
Comment (mgaffney) [inline internal/auth/oidc/repository_auth_method_create.go:12]:
```suggestion
// CreateAuthMethod creates am (*AuthMethod) in the repo along with its
```

File: internal/auth/oidc/repository_auth_method_create.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+// associated embedded optional value objects of SigningAlgs, CallbackUrls,
+// AudClaims and Certificates. and returns the newly created AuthMethod
```
Comment (mgaffney) [inline internal/auth/oidc/repository_auth_method_create.go:14]:
```suggestion
// AudClaims, and Certificates and returns the newly created AuthMethod
```

**Review by mgaffney — APPROVED** [review]
Just some minor comments

File: internal/auth/oidc/provider.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+// setProvider will set an entry in the cache.
+func (c *providers) setProvider(ctx context.Context, authMethodId string, p *oidc.Provider) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cache[authMethodId] = p
+}
+
+// delProvider will delete an entry in the cache.
+func (c *providers) delProvider(ctx context.Context, authMethodId string) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/auth/oidc/provider.go:93]: ty!

File: internal/auth/oidc/provider.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+		default:
+			c.delProvider(ctx, currentFromDb.PublicId)
+		}
+	}
+	c.setProvider(ctx, currentFromDb.PublicId, storedProvider)
+	return storedProvider, nil
+}
+
+// setProvider will set an entry in the cache.
+func (c *providers) setProvider(ctx context.Context, authMethodId string, p *oidc.Provider) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/auth/oidc/provider.go:86]: ty!

File: internal/auth/oidc/provider.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+//
+// Before returning a cached oidc.Provider, getProvider ensures that the
+// AuthMethod data used for the oidc.Provider's configuration hasn't changed
+// since it was cached. This is necessary because another controller could
+// update the AuthMethod in the DB, which would require changing the
+// configuration of the cached provider.
+//
+// getProvider will update the providerCache with the new AuthMethod, if it
+// determines the provider's configuration has been updated in the DB.
+func (c *providers) getProvider(ctx context.Context, currentFromDb *AuthMethod) (*oidc.Provider, error) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/auth/oidc/provider.go:56]: I'm going to keep the ctx... loggers, metrics, tracing, tigers and bears are coming soon.

File: internal/auth/oidc/repository_auth_method_delete_test.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+				databaseWrapper, err := kmsCache.GetWrapper(context.Background(), org.PublicId, kms.KeyPurposeDatabase)
+				require.NoError(t, err)
+				return TestAuthMethod(t, conn, databaseWrapper, org.PublicId, InactiveState, TestConvertToUrls(t, "https://alice.com")[0], "alice_rp", "alices-dogs-name")
+			}(),
+			wantRowsDeleted: 1,
+		},
+		{
+			name:         "no-public-id",
+			authMethod:   func() *AuthMethod { am := AllocAuthMethod(); return &am }(),
+			wantErrMatch: errors.T(errors.InvalidPublicId),
```
Comment (s-christoff) [inline internal/auth/oidc/repository_auth_method_delete_test.go:43]: What is the reason for not adding WantRowsDeleted: 0 here?

**Review by s-christoff — APPROVED** [review]
🤙

File: internal/auth/oidc/repository_auth_method_delete_test.go (hunk 13 above)
Comment (jimlambrt) [reply internal/auth/oidc/repository_auth_method_delete_test.go:43]: I'm not always consistent, but I try not to add zero values explicitly for tests.  Maybe that's not the best choice?

Comment (louisruch) [reply internal/auth/oidc/repository_auth_method_delete_test.go:43]: I am a fan of not explicitly adding zero values

**Review by mgaffney — APPROVED** [review]

File: internal/auth/oidc/repository_auth_method_update.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+// If the AuthMethod contains a DiscoveryUrl for an OIDC provider, TestAuthMethod
+// retrieves the OpenID Configuration document. The values in the AuthMethod
+// (and associated data) are validated with the retrieved document. The issuer and
+// id token signing algorithm in the configuration are validated with the
+// retrieved document. TestAuthMethod also verifies the authorization, token,
+// and user_info endpoints by connecting to each and uses any certificates in the
+// configuration as trust anchors to confirm connectivity.
+//
+// Options supported are: WithPublicId, WithAuthMethod
+func (r *Repository) TestAuthMethod(ctx context.Context, opt ...Option) error {
```
Comment (talanknight) [inline internal/auth/oidc/repository_auth_method_update.go:46]: Consider renaming ValidateAuthMethod given the existing meaning of TestX in go and the collision with the function in testing.go :)

**Review by talanknight — APPROVED** [review]

File: internal/auth/oidc/repository_auth_method_update.go (hunk 14 above)
Comment (jimlambrt) [reply internal/auth/oidc/repository_auth_method_update.go:46]: Ty!  I've just bumped into this while implementing unit tests.  I'll change the name in the next PR.

**Pull request merged**